examples/Example11/electrons.cu

```diff
@@ -0,0 +1,255 @@
+#include <G4HepEmInteractionUtils.icc>
+#include <G4HepEmElectronManager.icc>
+#include <G4HepEmElectronInteractionBrem.icc>
+#include <G4HepEmElectronInteractionIoni.icc>
+#include <G4HepEmPositronInteractionAnnihilation.icc>
+
+// Compute the physics and geometry step limit, transport the electrons while
+// applying the continuous effects and maybe a discrete process that could
+// generate secondaries.
+template <bool IsElectron>
+static __device__ __forceinline__ void TransportElectrons(Track *electrons, const adept::MParray *active,
+                                                          Secondaries &secondaries, adept::MParray *activeQueue,
+                                                          GlobalScoring *scoring)
+{
+  constexpr int Charge  = IsElectron ? -1 : 1;
+  constexpr double Mass = copcore::units::kElectronMassC2;
+  fieldPropagatorConstBz fieldPropagatorBz(BzFieldValue);
+
+  int activeSize = active->size();
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < activeSize; i += blockDim.x * gridDim.x) {
+    const int slot      = (*active)[i];
+    Track &currentTrack = electrons[slot];
+
+    // Init a track with the needed data to call into G4HepEm.
+    G4HepEmElectronTrack elTrack;
+    G4HepEmTrack *theTrack = elTrack.GetTrack();
+    theTrack->SetEKin(currentTrack.energy);
+    // For now, just assume a single material.
+    int theMCIndex = 1;
+    theTrack->SetMCIndex(theMCIndex);
+    theTrack->SetCharge(Charge);
+
+    // Sample the `number-of-interaction-left` and put it into the track.
+    for (int ip = 0; ip < 3; ++ip) {
+      double numIALeft = currentTrack.numIALeft[ip];
+      if (numIALeft <= 0) {
+        numIALeft                  = -std::log(currentTrack.Uniform());
+        currentTrack.numIALeft[ip] = numIALeft;
+      }
+      theTrack->SetNumIALeft(numIALeft, ip);
+    }
+
+    // Call G4HepEm to compute the physics step limit.
+    G4HepEmElectronManager::HowFar(&g4HepEmData, &g4HepEmPars, &elTrack);
+
+    // Get result into variables.
+    double geometricalStepLengthFromPhysics = theTrack->GetGStepLength();
+    // The phyiscal step length is the amount that the particle experiences
+    // which might be longer than the geometrical step length due to MSC. As
+    // long as we call PerformContinuous in the same kernel we don't need to
+    // care, but we need to make this available when splitting the operations.
+    // double physicalStepLength = elTrack.GetPStepLength();
+    int winnerProcessIndex = theTrack->GetWinnerProcessIndex();
+    // Leave the range and MFP inside the G4HepEmTrack. If we split kernels, we
+    // also need to carry them over!
+
+    // Check if there's a volume boundary in between.
+    double geometryStepLength = fieldPropagatorBz.ComputeStepAndPropagatedState</*Relocate=*/false>(
+        currentTrack.energy, Mass, Charge, geometricalStepLengthFromPhysics, currentTrack.pos, currentTrack.dir,
+        currentTrack.currentState, currentTrack.nextState);
```

Comment:
Ah, I didn't realize this, indeed. I will see what I can do, maybe I will add a separate propagator within Example11 to avoid changing this in all examples.

Reply:
Yeah, swapping out the `LoopNavigator` from the propagator has a more noticeable effect :-)
```
epsftws build $ diff -u example9.log example11.log 
--- example9.log	2021-06-08 12:35:39.342576794 +0200
+++ example11.log	2021-06-08 12:35:28.382581939 +0200
@@ -233,4 +233,4 @@
 iter  212 -- tracks in flight:     1 energy deposition:    88.4433 number of secondaries: 62973 number of hits: 67728
 iter  213 -- tracks in flight:     2 energy deposition:    88.4433 number of secondaries: 62974 number of hits: 67728
 iter  214 -- tracks in flight:     0 energy deposition:    88.4437 number of secondaries: 62974 number of hits: 67728
-Run time: 0.8873
+Run time: 0.0534
epsftws build $ BuildProducts/bin/example9 -gdml_name /srv/geant4/gdml/cms2018.gdml > example9-cms.log
epsftws build $ BuildProducts/bin/example11 -gdml_name /srv/geant4/gdml/cms2018.gdml > example11-cms.log
epsftws build $ grep ^Run *-cms.log
example11-cms.log:Run time: 0.3014
example9-cms.log:Run time: 0.5901
```